### ISSUE TITLE
BUG: DatetimeIndex intersection is empty or garbage

### ISSUE THREAD

**Issue by yportier** [pandas-dev/pandas#59271]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

index = pd.date_range('2024-01-01', '2024-02-19', unit='s')
period = pd.date_range('2024-01-01', '2024-02-09', unit='s')
index.intersection(period)  # ughh... DatetimeIndex(['2024-01-01', '2739931-01-04'], dtype='datetime64[s]', freq='D')

index = pd.date_range('2024-01-01', '2024-02-19', unit='s')
period = pd.date_range('2024-01-09', '2024-02-09', unit='s')

index.intersection(period)  # empty!
```

### Issue Description

In the first example, the result is obviously buggy with year 2739931
The second example is empty when it shouldn't

### Expected Behavior

both examples are expected to return an index equivalent to `period`

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.10.12.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 6.1.85+
Version               : #1 SMP PREEMPT_DYNAMIC Thu Jun 27 21:05:47 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : en_US.UTF-8
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.2
numpy                 : 1.25.2
pytz                  : 2023.4
dateutil              : 2.8.2
setuptools            : 67.7.2
pip                   : 23.1.2
Cython                : 3.0.10
pytest                : 7.4.4
hypothesis            : None
sphinx                : 5.0.2
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : 4.9.4
html5lib              : 1.1
pymysql               : None
psycopg2              : 2.9.9
jinja2                : 3.1.4
IPython               : 7.34.0
pandas_datareader     : 0.10.0
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2023.6.0
gcsfs                 : 2023.6.0
matplotlib            : 3.7.1
numba                 : 0.58.1
numexpr               : 2.10.1
odfpy                 : None
openpyxl              : 3.1.5
pandas_gbq            : 0.19.2
pyarrow               : 14.0.2
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.11.4
sqlalchemy            : 2.0.31
tables                : 3.8.0
tabulate              : 0.9.0
xarray                : 2023.7.0
xlrd                  : 2.0.1
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

**Comment by Animesh-Shukla** [comment]
take

**Comment by yportier** [comment]
Apologies, I did not install the main branch correctly, I confirm the bug is not present in 3.0.0.dev0+1223.gc3b72aae6f

**Comment by asishm** [comment]
Thanks! This was fixed in #59037 and will be released in 3.0
